### PR TITLE
Save 600+ bytes RAM (step 4 of 9) - deduplicate scheduler ISR functions

### PR DESCRIPTION
**Pull request by adbancroft**
This is the 4th PR in a series of scheduler focused PRs that end up saving >600 bytes RAM (compared to master as of 2023-06-11)

This PR removes the duplicate code in the per-channel scheduler ISR functions, routing them through an equivalent function that handles all channels:
```
ignitionScheduleISR
fuelScheduleISR
```
These are completely inlined, so no loss in performance.

The end game is [this branch](https://github.com/adbancroft/speeduino/tree/SCHD_8_scheduler_arrays). Since there are so many changes, I split them into as a series of PRs (each with a specific focus), to make the changes easier to understand, test and get feedback on.

1. https://github.com/noisymime/speeduino/pull/1010 (merged)
2. https://github.com/noisymime/speeduino/pull/1039 (merged)
3. #1018 
4. This PR
5. De-duplicate setSchedule functions
6. De-duplicate schedule structs
7. Optimize ignition schedule struct
8. Optimize fuel schedule struct
9. Use arrays to store schedulers.

### CODE REVIEW

File: speeduino/timers.ino [hunk 1]
```diff
@@ -63,14 +85,20 @@ void oneMSInterval(void) //Most ARM chips can simply call a function
...
-  if(ignitionSchedule5.Status == RUNNING) { if( (ignitionSchedule5.startTime < targetOverdwellTime) && (configPage4.useDwellLim) && (isCrankLocked != true) ) { ign5EndFunction(); ignitionSchedule5.Status = OFF; } }
-  if(ignitionSchedule6.Status == RUNNING) { if( (ignitionSchedule6.startTime < targetOverdwellTime) && (configPage4.useDwellLim) && (isCrankLocked != true) ) { ign6EndFunction(); ignitionSchedule6.Status = OFF; } }
-  if(ignitionSchedule7.Status == RUNNING) { if( (ignitionSchedule7.startTime < targetOverdwellTime) && (configPage4.useDwellLim) && (isCrankLocked != true) ) { ign7EndFunction(); ignitionSchedule7.Status = OFF; } }
-  if(ignitionSchedule8.Status == RUNNING) { if( (ignitionSchedule8.startTime < targetOverdwellTime) && (configPage4.useDwellLim) && (isCrankLocked != true) ) { ign8EndFunction(); ignitionSchedule8.Status = OFF; } }
+  if(ignitionSchedule1.Status == RUNNING) { if( (ignitionSchedule1.startTime < targetOverdwellTime) && (configPage4.useDwellLim) && (isCrankLocked != true) ) { ignitionSchedule1.pEndCallback(); ignitionSchedule1.Status = OFF; } }
```
Comment (VitorBoss) [inline speeduino/timers.ino:88]: ``  if(ignitionSchedule1.Status == RUNNING) { if( (ignitionSchedule1.startTime < targetOverdwellTime) && (configPage4.useDwellLim) && (isCrankLocked != true) ) { ignitionSchedule1.pEndCallback(); ignitionSchedule1.Status = OFF; } }
``
Please don't consider bit fields as bools, please change it to `(configPage4.useDwellLim == 1)` and as the C compiled code runs each check in sequence is better to use the limiter first like this:
``  if ((configPage4.useDwellLim == 1) && (isCrankLocked != true) && (ignitionSchedule1.Status == RUNNING) && (ignitionSchedule1.startTime < targetOverdwellTime)) { ignitionSchedule1.pEndCallback(); ignitionSchedule1.Status = OFF; }
``
This way if dwell limit is turned off the execution don't expend too much time, not critical, but is optimal, also avoiding nested `if` statements are better for clarity, I know this is from old code, but is better to fix it in a PR right? 😉

Comment (adbancroft) [reply speeduino/timers.ino:88]: > I know this is from old code, but is better to fix it in a PR right?

[Boy scout rule](https://matheus.ro/2017/12/11/clean-code-boy-scout-rule/) :wink:

If I make that change, I may as well hoist the the config check into an outer ` if ((configPage4.useDwellLim == 1) && (isCrankLocked != true) { /* Per channel checks here */ }`.

Comment (VitorBoss) [reply speeduino/timers.ino:88]: Even better, the current way do the most important checks for last.

Comment (adbancroft) [reply speeduino/timers.ino:88]: And if doing that, I may as well encapsulate the per-channel check into an inline method:
```
static inline void applyOverDwellCheck(IgnitionSchedule &schedule, uint32_t targetOverdwellTime) {
  if ((schedule.Status == RUNNING) && (schedule.startTime < targetOverdwellTime)) { 
    schedule.pEndCallback(); schedule.Status = OFF; 
  }
}
```

File: test/test_schedules/test_status_initial_off.cpp [hunk 2]
```diff
@@ -31,32 +30,40 @@ void test_status_initial_off_inj4(void)
 #if INJ_CHANNELS >= 5
 void test_status_initial_off_inj5(void)
 {
+#if ING_CHANNELS >= 5
```
Comment (noisymime) [inline test/test_schedules/test_status_initial_off.cpp:33]: There's a typo in `ING_CHANNELS` here. I don't believe that `#if` is required anyway as there's already one around the coverall function declaration. 

Same applies for all of these >=5

Comment (adbancroft) [reply test/test_schedules/test_status_initial_off.cpp]: Fixed - and addressed a tennsy unit test build issue.

**Comment by noisymime** [comment]
@adbancroft I've gone through and merged all the big stuff from master into this and have done a few basic checks. 

Are you confident for this getting pulled in in its current form?

**Comment by adbancroft** [comment]
> Are you confident for this getting pulled in in its current form?

I reviewed all code changes between the PR & master. Made 1 code change, plus enabled Teensy35 unit tests.

All unit tests pass on Mega & Teensy35. I bench tested various combinations of ign & inj up to 4 channels with a logic analyzer.

I'm as confident as I can be, given the ino->cpp changes.

**Pull request merged**